### PR TITLE
Refactoring equivalent atom and parametrizable dihedral detection

### LINKED ISSUE #422: Parametrize: incorrect equivalent atom detection
For this nice molecule (red -- oxygen, yellow -- sulphur):
mol
[mol.mol2.txt](https://github.com/Acellera/htmd/files/1204442/mol.mol2.txt)

Then running *parametrize*:
```bash
parametrize -m mol.mo2
```

It reports that there are two identical carbons:
```
...
Equivalent atom groups:
 O1
 C1 C2
 S1
 O2
 S2
 H1
 H2
...
```

What is obviously incorrect!

### LINKED ISSUE #485: Parameterizable angle detection depends on atom order
Two identical structure of chlorethene, except the 3rd and 4th atoms are swapped:
```
$ cat chlorethene_1.mol2
@<TRIPOS>MOLECULE
*****
 6 0 0 0 0
SMALL
GASTEIGER

@<TRIPOS>ATOM
      1 C           0.9437    0.0452   -0.0669 C.2     1  UNL1       -0.0876
      2 C           2.2765    0.0400   -0.0741 C.2     1  UNL1       -0.0025
      3 CL          3.1933   -0.8085   -1.2443 Cl      1  UNL1       -0.0925
      4 H           2.8937    0.5621    0.6459 H       1  UNL1        0.0740
      5 H           0.3971    0.5950    0.6913 H       1  UNL1        0.0543
      6 H           0.3695   -0.4954   -0.8124 H       1  UNL1        0.0543
@<TRIPOS>BOND
```
```
$ cat chlorethene_2.mol2
@<TRIPOS>MOLECULE
*****
 6 0 0 0 0
SMALL
GASTEIGER

@<TRIPOS>ATOM
      1 C           0.9437    0.0452   -0.0669 C.2     1  UNL1       -0.0876
      2 C           2.2765    0.0400   -0.0741 C.2     1  UNL1       -0.0025
      3 H           2.8937    0.5621    0.6459 H       1  UNL1        0.0740
      4 CL          3.1933   -0.8085   -1.2443 Cl      1  UNL1       -0.0925
      5 H           0.3971    0.5950    0.6913 H       1  UNL1        0.0543
      6 H           0.3695   -0.4954   -0.8124 H       1  UNL1        0.0543
@<TRIPOS>BOND
```

Try to list parameterizable dihedral angles:
```
$ parameterize chlorethene_1.mol2 --list
 === Parameterizable dihedral angles of chlorethene_1.mol2 ===

  CL1-C2-C1-H
```
```
$ parameterize chlorethene_2.mol2 --list
 === Parameterizable dihedral angles of chlorethene_1.mol2 ===

  H1-C2-C1-H2
```

`Cl-C-C-H` and `H-C-C-H` are different angles!

### PR DESCRIPTION
**Pull request by raimis**
- [x] Clean
- [x] Improve equivalent atom detection
  - [x] NetworkX-based algorithm
  - [x] Fix #422
  - [x] Tests
- [x] Improve parametrizable dihedral angle detection
  - [x] NetworkX-based algorithm
  - [x] Integrate
  - [x] Fix #485
  - [x] Skip triple bonds
  - [x] Update tests
- [x] Documentation

### CODE REVIEW

**Comment by raimis** [comment]
Fish test data update then #521 is merged!

File: htmd/parameterization/detect.py [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+
+    return nx.is_isomorphic(graph1, graph2, node_match = lambda node1, node2: node1['element'] == node2['element'])
+
+def detectEquivalentAtoms(molecule):
+    """
+    Detect topologically equivalent atoms.
+
+    Arguments
+    ---------
+    molecule : FFMolecule
```
Comment (stefdoerr) [inline htmd/parameterization/detect.py]: `:class:\`FFMolecule <htmd.ffmole.blabla.FFMolecule>\`` for the online docs to work

**Comment by raimis** [comment]
@stefdoerr when could you finish review this?

File: htmd/parameterization/detect.py [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+    methyl.add_edge(0, 2)
+    methyl.add_edge(0, 3)
+
+    return methyl
+
+def detectParameterizableCores(graph):
+    """
+    Detect parametrizable dihedral angle cores (central atom pairs)
+
+    The cores are detected by looking for bridges (bonds with divide the molecule into two parts) in a molecular graph.
```
Comment (stefdoerr) [inline htmd/parameterization/detect.py]: `with` = `which`

File: htmd/parameterization/detect.py [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+    2. Largest atomic number weighted closeness centrality
+    """
+
+    terminals = list(sideGraph.neighbors(centre))
+
+    # Get a subgraph for each terminal
+    sideGraph = sideGraph.copy()
+    sideGraph.remove_node(centre)
+    terminalGraphs = list(nx.connected_component_subgraphs(sideGraph))
+    terminalGraphs = [[terminalGraph for terminalGraph in terminalGraphs if terminal in terminalGraph.nodes][0] for terminal in terminals]
```
Comment (stefdoerr) [inline htmd/parameterization/detect.py]: Rewrite this please as a double loop. Nested list comprehensions are hard to read

File: htmd/parameterization/detect.py [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+    weightedCentralities = [weighted_closeness_centrality(graph, terminal, weight='number') for terminal in terminals]
+    scores = list(zip(centralities, weightedCentralities))
+
+    # Choose the terminals
+    chosen_terminals = []
+    refTerminalGraph = None
+    for terminal, score, terminalGraph in zip(terminals, scores, terminalGraphs):
+        if score < max(scores):
+            continue
+
```
Comment (stefdoerr) [inline htmd/parameterization/detect.py]: So you don't really want a loop. Do an `argmax` on `scores` and index `terminals` and `terminalGraphs` with that value.

**Comment by stefdoerr** [comment]
In addition to the above comments:
1. The unrolling of cyclic graphs into trees depends on the order of traversal of the neighbours no? 
So if the atom order changes in the molecule you might not recognize them as isomorphic trees.
2. Just a minor thingy but you could name private functions of the module which you don't want users to access from outside the module with `_` like `_getMethylGraph`.

**Comment by j3mdamas** [comment]
Wow, this seems like an impressive PR! Good luck! Let me know if you need any help.

File: htmd/parameterization/detect.py (hunk 4 above)
Comment (raimis) [reply htmd/parameterization/detect.py]: Usually there is more than one terminal with the same score, so `argmax` isn't a replacement here.

File: htmd/parameterization/detect.py (hunk 3 above)
Comment (raimis) [reply htmd/parameterization/detect.py]: Rewrote as a single-loop list comprehension... :D

File: htmd/parameterization/detect.py (hunk 1 above)
Comment (raimis) [reply htmd/parameterization/detect.py]: Done!

File: htmd/parameterization/detect.py (hunk 2 above)
Comment (raimis) [reply htmd/parameterization/detect.py]: Done!

**Comment by raimis** [comment]
@stefdoerr 

1. `nx.is_isomorphic` takes into account that.
2. Done!

**Comment by stefdoerr** [comment]
Got some conflicts to resolve

**Comment by raimis** [comment]
@stefdoerr the conflicts are resolved! Any more comments?

**Comment by stefdoerr** [comment]
Nah

**Pull request merged**